clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analysis_service.py

```diff
@@ -1,37 +1,107 @@
+        elif fam_ct == 0 and hex_ct > self.HEX_THRESHOLD:
+            # TODO Should retest and failed be the same thing or not.
+            return COVID_RESPONSE_FAILED
+        elif fam_ct <= self.FAM_THRESHOLD:
+            return COVID_RESPONSE_POSITIVE
+        else:
+            # TODO Don't know if it is actually a good idea to raise here or not...
+            raise NotImplementedError(
+                "Got CT-value for FAM: {} and CT-value for HEX: {}.".format(fam_ct, hex_ct))
 
-    def validate_control_value(self, control_type, ct_value, lower_bound, upper_bound):
-        if control_type.lower() == RT_PCR_POSITIVE_CONTROL and ct_value <= lower_bound:
-            self.has_valid_pos_control = True
-        if control_type.lower() == RT_PCR_NEGATIVE_CONTROL and ct_value >= upper_bound:
-            self.has_valid_neg_control = True
+    def _analyze_controls(self, positive_controls, negative_controls):
+        errors = []
+        for pos_control in positive_controls:
+            if self._analyze_sample(pos_control) == COVID_RESPONSE_NEGATIVE:
+                errors += PositiveControlWasNegative(
+                    "Positive control sample: {} was negative for covid-19".format(pos_control["id"]))
 
-    def is_valid(self):
-        return self.has_valid_neg_control and self.has_valid_pos_control
+        for neg_control in negative_controls:
+            if self._analyze_sample(neg_control) == COVID_RESPONSE_POSITIVE:
+                errors += NegativeControlWasPositive(
+                    "Negative control sample: {} was positive for covid-19".format(
+                        neg_control["id"]))
 
-    def analyze(self, ct_value, lower_bound, upper_bound):
+        if errors:
+            for e in errors:
+                log.error(e.message)
+            raise MultipleAnalysisErrors(errors)
+
+    def analyze_samples(self, positive_controls, negative_controls, samples):
         """
-        Returns: String with value of either 'positive', 'negative', 'failed'
+        This assumes all controls and samples are from the same plate.
+        All samples, and controls should be submitted as dict-like objects on the format:
+         {"id": "<sample id>", "FAM-CT": <value as numeric>, "HEX-Ct": <value as numric>}
+
+        The method will return a generator of objects on the format:
+         {"id": "<same as sample id above>",
+             "diagnosis_result": "<CONST COVID RESPONSE>"}
         """
-        if not self.has_valid_pos_control or not self.has_valid_neg_control:
-            raise UsageError("There are no valid control samples")
-        if ct_value <= lower_bound:
-            return 'positive'
-        elif ct_value >= upper_bound:
-            return 'negative'
-        return 'failed'
+
+        # TODO Should we validate that there are controls on the plate?
+
+        # If controls fail, fail all samples on plate.
+        try:
+            self._analyze_controls(positive_controls, negative_controls)
```

Comment:
I wonder if we should ask whether there should be different actions for failed controls and failed individual samples? Perhaps failed controls means that the plate should be re-runned, but a failed individual samples goes on in workflow and reported as failed? It seems to me that a control failure could not be so easily distinguished from sample failure.

Reply:
Ask Maike if we should separate different failure states, i.e. `FAILED_FROM_CONTROLS`, `FAILED_FROM_SAMPLE`.